build-scripts/build_xccdf.py

```diff
@@ -143,12 +115,23 @@ def main():
     loader.load_benchmark(benchmark_root)
 
     loader.add_fixes_to_rules()
+    xccdftree = loader.export_benchmark_to_xml()
 
-    if args.thin_ds_components_dir != "off":
-        link_benchmark_per_profile(loader, args)
+    checks = xccdftree.findall(".//{%s}check" % ssg.constants.XCCDF12_NS)
 
-    xccdftree = loader.export_benchmark_to_xml()
-    link_benchmark(loader, xccdftree, args)
+    oval_linker = link_oval(xccdftree, checks, args.oval, args.build_ovals_dir)
+
+    ocil = loader.export_ocil_to_xml()
+    link_ocil(xccdftree, checks, args.ocil, ocil)
+
+    ssg.xml.ElementTree.ElementTree(xccdftree).write(args.xccdf)
+
+    if args.thin_ds_components_dir != "off":
+        if not os.path.exists(args.thin_ds_components_dir):
+            os.makedirs(args.thin_ds_components_dir)
+        store_xccdf_per_profile(loader, args.thin_ds_components_dir)
```

Comment:
When I build thin data streams for RHEL 9 product, it tracebacks here:

```
[ 30%] [rhel9-content] generating plain XCCDF, OVAL and OCIL files
Traceback (most recent call last):
  File "/home/jcerny/work/git/content/build-scripts/build_xccdf.py", line 138, in <module>
    main()
  File "/home/jcerny/work/git/content/build-scripts/build_xccdf.py", line 132, in main
    store_xccdf_per_profile(loader, args.thin_ds_components_dir)
  File "/home/jcerny/work/git/content/build-scripts/build_xccdf.py", line 93, in store_xccdf_per_profile
    for id_, xccdftree in loader.get_benchmark_xml_by_profile():
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 1603, in get_benchmark_xml_by_profile
    profile_id, benchmark = self.benchmark.get_benchmark_xml_for_profile(profile)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 518, in get_benchmark_xml_for_profile
    return profile.id_, self.to_xml_element(
                        ^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 469, in to_xml_element
    self._add_groups_xml(root, components_to_not_include, env_yaml)
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 439, in _add_groups_xml
    root.append(group.to_xml_element(env_yaml, components_to_not_include))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 723, in to_xml_element
    self._add_sub_groups(group, components_to_not_include, env_yaml)
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 693, in _add_sub_groups
    group.append(_group.to_xml_element(env_yaml, components_to_not_include))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 723, in to_xml_element
    self._add_sub_groups(group, components_to_not_include, env_yaml)
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 693, in _add_sub_groups
    group.append(_group.to_xml_element(env_yaml, components_to_not_include))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 723, in to_xml_element
    self._add_sub_groups(group, components_to_not_include, env_yaml)
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 693, in _add_sub_groups
    group.append(_group.to_xml_element(env_yaml, components_to_not_include))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 722, in to_xml_element
    self._add_rules_xml(group, rules_to_not_include, env_yaml)
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 648, in _add_rules_xml
    group.append(rule.to_xml_element(env_yaml))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 1132, in to_xml_element
    add_reference_elements(rule, self.references, ref_uri_dict)
  File "/home/jcerny/work/git/content/ssg/build_yaml.py", line 120, in add_reference_elements
    raise ValueError(msg)
ValueError: Error processing reference cis: ['5.6.1.4']. A reference type has been added that the project doesn't know about.
```

The problem is related to the references, specifically, how reference URIs are  stored and processed throughout the built process. Unfortunately, we now have product-specific reference types and global reference types. Global reference types are defined in `ssg/constants.py` and the product-specific ones are set in each product's `product.yml`. The reference URIs are included into the infamous variable `env_yaml`. See the `Rule.to_xml_element()` method in  `build_yaml.py` around line 1128. I think we probably need to pass `env_yaml` to this function and then pass it down to the called functions, as can we infer from the traceback.

Reply:
`fd product.yml products | xargs grep cis`

Reply:
I fixed traceback.